roachprod: roachtest: refactoring

smg260 opened the pull request:
This commit addresses some usability issues within roachprod/roachtest, and does some minor clean up.

Summary:

`roachprod`

- `c.Parallel` now accepts  `nodes Nodes` and `func(ctx context.Context, node Node)`. Previously, these was an `int n` representing the first `n` nodes to operate on, and `func(ctx context.Context, int n)`, where second arg would confusingly represented the index of the node being operated on, which was wholly dependent on a captured-by-reference slice at the call site.  This change makes it explicit to the caller and the function, which exact node is targeted. 

- `c.Parallel` now returns `[]*RunResultDetails, boolean, error`. The slice will contain results, successful and failed, for each node (or thus far, depending on fail-fast behaviour). The boolean is a signal to the caller that at least one of the results in the slice suffered from a command error, which can be inspected in the result itself. The `error` now more idiomatically represents an unrecoverable error encountered during `Parallel`; i.e. an un retryable error in `roachprod`. In the latter case, results will be `nil`.

- All calls to execute a remote command on a node are all funnelled to `c.runCmdOnSingleNode`, ensuring consistency, and reducing duplicated code. This function accepts a `RunCmdOptions`  struct to configure it's behaviour, including `stdin/out`

- `runCmdOnSingleNode` : in error cases, include stdout/err output (truncated after n bytes - effectively `tail`). This output will show in the `failure_n.log` and give convenient insight. Full output is always retained in the run log.

- `RunResultDetails.{Stdout,Stderr,CombinedOut}` are now all `string` to represent the most common use case, and a convenience `r.Output()` added to print out the contents of `Stderr,Stdout,CombinedOut`

- Remove complex legacy pipe processing logic from `roachtest` (`execCmd,execCmdEx`), previously used when `roachprod` was not a library of `roachtest`

- When showing the result of a command on each node on the CLI, explicitly display `<ok>` or `<err>` followed by output/error


`roachtest`

- skip post-test assertions on a test failure

- removal of `execCmd/execCmdRes` which were doing some complex pipe processing before calling roachprod.Run (not required since roachprod is a lib of roachtest)

Minor

- Clean up various redundant/unused variables
- Typos
- Command formatting

Informs: #104312 
Informs: #104316 (TBD)

Epic: none
Release note: none

Commit:
roachprod: reuse `c.runCmdOnSingleNode` and refactor `c.Parallel`

- Funnel all ssh command calls to `runCmdOnSingleNode`. A boolean is added
to control whether we prepend ROACHPROD env vars, which are required for
user-supplied commands via `c.Run/RunWithDetails`

- `c.ParallelE` now accepts a node list to make it clear which nodes a
specified command is running on. Previously, and confusingly, this was
achieved via an `int` arg and a captured-by-ref slice of nodes. The int
arg referred to the first `int` nodes in the captured slice.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104915)
<!-- Reviewable:end -->

Commit:
roachtest: skip post-test assertions on failure

Commit:
roachtest: roachprod: cleanup, usability

roachtest
- Remove `execCmd/Ex` pipes and buffers, and instead dispatch to `roachprod.Run` directly.
- `createFailedFile` fn checks for name length
- `RunE` wraps `err` with location of full output
- `c.RunWithDetails` creates a failed file if any node returns an error
 (previously just for a roachprod error)
- `test.log` shows where a command is logging to
- `c.runCmdOnSingleNode` includes command and output in the error
- fix `fileExistsOnFirstNode` (was returning true if file not exists instead of vice versa)
- update `c.{startNode,initializeCluster,setClusterSettings}` to return (*RunResultDetails, error)
- simplify `c.Start`

roachprod
- `RunResultDetails.Output()` convenience to print cmd output
- `c.runCmdOnSingleNode` includes truncated stdout/err in error object;
the full command output will still be available in the cmd run log.
- `processResults` formatting; use tabs, show `<ok>` or `<err>` prefix instead of blank output

smg260 commented:
[TeamCity @ SELECT_PROBABILITY=0.3](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/11073200)

renatolabs reviewed (commented):
Definitely a lot of changes and it's hard to digest it all in one sitting. Overall I do quite like the changes though, just left minor comments.

I'm looking at the run on this branch that you linked above. I was curious about the new formatting of the command output, but couldn't find anything. Some things also caught my attention and I have the feeling they are not expected:

* the `.failed` log file seems to be missing.
* The `jepsen` failure has an empty log file for the `jepsen` command.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-:-NarpC-95Rj4vX3qcGwN:b-d3ar5y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @smg260)

___
*[`pkg/cmd/roachtest/cluster.go` line 589 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-NarcyXG8NJwMDUJxY6Y:-NarcyXG8NJwMDUJxY6Z:b-w4tk4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bbf620c9ad337cedf6c07e94c95d2b27cfcbf4c/pkg/cmd/roachtest/cluster.go#L589)):*
> ```Go
> 		err:    err,
> 		stdout: stdoutString,
> 		stderr: stderrString,
> ```

Nice, :fire: 
___
*[`pkg/cmd/roachtest/cluster.go` line 2207 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-NareDUSERNJedMD9h0N:-NareDUSERNJedMD9h0O:b-vlub7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/cmd/roachtest/cluster.go#L2207)):*
> ```Go
> 	logFileFull := l.File.Name()
> 	if err != nil {
> 		createFailedFile(logFileFull)
> ```

In `RunE`, we call `createFailedFile(logFile)`. Any reason this needs to be `createFailedFile(logFileFull)` in this function?
___
*[`pkg/cmd/roachtest/cluster.go` line 2226 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-Nareiyj2Y5kbTMwmfTw:-Nareiyj2Y5kbTMwmfTx:bwon8v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/cmd/roachtest/cluster.go#L2226)):*
> ```Go
> func createFailedFile(logFile string) {
> 	if logFile == "" {
> 		return
> ```

Why the check? Can this actually be empty in any of the callers?
___
*[`pkg/roachprod/roachprod.go` line 1760 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-Narf8My36PyalBNZeUX:-Narf8My36PyalBNZeUY:b-pmz28y) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/roachprod/roachprod.go#L1760)):*
> ```Go
> 		res := &install.RunResultDetails{Node: node}
> 
> 		//TODO: this usage may not always be correct, if the target nodes are not sequential
> ```

The current implementation of `TargetNodes` already guarantees ordering. I'd either remove the `TODO` to avoid confusion (same for TODO below)
___
*[`pkg/roachprod/install/cluster_synced.go` line 876 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-Narj0w6-kfgM3vJLYWO:-Narj0w6-kfgM3vJLYWP:b-knvnqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/roachprod/install/cluster_synced.go#L876)):*
> ```Go
> 		res = newRunResultDetails(node, cmdErr)
> 		res.CombinedOut = string(out)
> 		//output = fmtOut(res.CombinedOut)
> ```

Can be removed?
___
*[`pkg/roachprod/install/cluster_synced.go` line 1134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-NarkNHA3h12UykBX2Ei:-NarkNHA3h12UykBX2Ej:bhofei7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/roachprod/install/cluster_synced.go#L1134)):*
> ```Go
> 		runOpts.combinedOut = false
> 		return c.runCmdOnSingleNode(ctx, l, n, cmd, runOpts)
> 		//sess := c.newSession(l, 1, cmd, withDebugName("ssh-gen-key"))
> ```

Delete?
___
*[`pkg/roachprod/install/cluster_synced.go` line 1148 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-NarkSqRAmox62V38BZh:-NarkSqRAmox62V38BZi:bhofehq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/roachprod/install/cluster_synced.go#L1148)):*
> ```Go
> 		runOpts.stdin = bytes.NewReader(sshTar)
> 		return c.runCmdOnSingleNode(ctx, l, node, `tar xf -`, runOpts)
> 		//sess := c.newSession(l, node, cmd, withDebugName("ssh-dist-key"))
> ```

Delete.
___
*[`pkg/roachprod/install/cluster_synced.go` line 2452 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-NarkxA353JwU6Md8C0F:-NarkxA353JwU6Md8C0G:b-5w76dq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/roachprod/install/cluster_synced.go#L2452)):*
> ```Go
> // on the cluster.
> //
> // ParallelE  runs at most `concurrency` (or  `config.MaxConcurrency` if it is lower) in parallel.
> ```

Nit: extra space.


<!-- Sent from Reviewable.io -->

smg260 reviewed (commented):
Thankyou for the review:

> the `.failed` log file seems to be missing.
Addressed in comments below

> The `jepsen` failure has an empty log file for the `jepsen` command.
The cmd log files show stdout/err only, and the jepsen command redirects both to invoke.log, hence no output. 

However, I've made an update to show some info in the command logs regardless:
- the full command being run
- the final status (<ok> or error)


I'll run another roachtest at 0.05 select to showcase the changes, and will update this thread when it's ready.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-:-NavzP4bCwrpGHnIIQM6:bb7eujr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan and @renatolabs)

___
*[`pkg/cmd/roachtest/cluster.go` line 2207 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-NareDUSERNJedMD9h0N:-NavgYUt6L8cbLqXaFtQ:bxkl7j3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/cmd/roachtest/cluster.go#L2207)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

In `RunE`, we call `createFailedFile(logFile)`. Any reason this needs to be `createFailedFile(logFileFull)` in this function?
</blockquote></details>

Good catch! Explains why there was no failed file for the `jepsen` test, but there was for `knex`. They should both be `logFileFull`
___
*[`pkg/cmd/roachtest/cluster.go` line 2226 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-Nareiyj2Y5kbTMwmfTw:-Navyqo49o8AoiMNsQUm:bqvwu19) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/cmd/roachtest/cluster.go#L2226)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Why the check? Can this actually be empty in any of the callers?
</blockquote></details>

I saw it occur at some point, but is now defensive.
___
*[`pkg/roachprod/roachprod.go` line 1760 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-Narf8My36PyalBNZeUX:-NavaHmr8_raEwlT5xzN:b-9w44m6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/roachprod/roachprod.go#L1760)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

The current implementation of `TargetNodes` already guarantees ordering. I'd either remove the `TODO` to avoid confusion (same for TODO below)
</blockquote></details>

You're right - there is a comment to that effect on `c.TargetNodes()` that I missed.
___
*[`pkg/roachprod/install/cluster_synced.go` line 876 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-Narj0w6-kfgM3vJLYWO:-NavahB87QUC-EB0Nydq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/roachprod/install/cluster_synced.go#L876)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Can be removed?
</blockquote></details>

Done.
___
*[`pkg/roachprod/install/cluster_synced.go` line 1134 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-NarkNHA3h12UykBX2Ei:-Navak_B5d-QxRNtGMc-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/roachprod/install/cluster_synced.go#L1134)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Delete?
</blockquote></details>

Done.
___
*[`pkg/roachprod/install/cluster_synced.go` line 1148 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-NarkSqRAmox62V38BZh:-NavalCj7we2xVSdqB1G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a7d06badb6f77513d3b9c47db976a62fc8f2927/pkg/roachprod/install/cluster_synced.go#L1148)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Delete.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

smg260 commented:
[Latest TC run](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/11185663?buildTab=tests)

**roachtest examples** 

_jepsen run log with no command output:_
<img width="1014" alt="image" src="https://github.com/cockroachdb/cockroach/assets/25202158/f16bea0b-430b-478b-aade-fdad41fafd61">

_failure_1.log with truncated output_
<img width="844" alt="image" src="https://github.com/cockroachdb/cockroach/assets/25202158/ff29e15c-e4eb-4e63-963c-732b34597d7d">
 
_excerpt of test.log showing first ~30chars of cmd being run, and where to find full details_
<img width="1155" alt="image" src="https://github.com/cockroachdb/cockroach/assets/25202158/195e94a6-bd4b-4557-9420-9aeda0a2d9ff">

_command log for command being run on multiple nodes_
<img width="652" alt="image" src="https://github.com/cockroachdb/cockroach/assets/25202158/60fe4eae-0528-4585-b881-08a48067b70d">

**roachprod cli examples**

_successful command on all nodes_
<img width="478" alt="image" src="https://github.com/cockroachdb/cockroach/assets/25202158/f9da75fd-40e8-42bb-bc19-9e35317ca722">

_successful command on 2/3 nodes_
<img width="532" alt="image" src="https://github.com/cockroachdb/cockroach/assets/25202158/51df3e1e-5431-45c7-ab87-d876a4a6338e">

renatolabs reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :tada: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-:-NbUdSPD-GIXlPzUlnao:by6fmpn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @herkolategan, @smg260, and @srosenberg)



<!-- Sent from Reviewable.io -->

herkolategan reviewed (approved):
> Reviewed 5 of 8 files at r4, 1 of 3 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104915)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/roachprod/roachprod.go` line 1080 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/104915#-Nb_6tdpEGDii8q14m9k:-Nb_6tdpEGDii8q14m9l:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c64af25d55d8ce568e5d9735ee74a87b9d3c32e/pkg/roachprod/roachprod.go#L1080)):*
> ```Go
> 	httpClient := httputil.NewClientWithTimeout(timeout)
> 	startTime := timeutil.Now().Unix()
> 	err = c.Parallel(ctx, l, c.TargetNodes(), func(ctx context.Context, node install.Node) (*install.RunResultDetails, error) {
> ```

Nice!


<!-- Sent from Reviewable.io -->

smg260 commented:
TFTR! 

I will hold off merging until I am completely back from PTO so I can handle any unexpected consequences.

smg260 commented:
bors r=renatolabs,herkolategan

smg260 commented:
blathers backport 23.1
